clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py

```diff
@@ -0,0 +1,650 @@
+    complete_data_df = complete_data_df.set_index(
+        ["participant_id", "session_id", "modality", "bids_filename"],
+        verify_integrity=True,
+    )
+    # open the reference for building the tsvs:
+    path_to_ref_csv = os.path.join(
+        os.path.dirname(os.path.dirname(os.path.dirname(__file__))),
+        "data",
+        "genfi_ref.csv",
+    )
+    df_ref = pd.read_csv(path_to_ref_csv, sep=";")
+    if not gif:
+        print(df_ref.head(8))
+    return {
+        col: complete_data_df.filter(items=list(df_ref[col]))
+        for col in ["participants", "sessions", "scans"]
+    }
+
+
+def intersect_data(
+    imaging_data: DataFrame, df_clinical_complete: DataFrame
+) -> DataFrame:
+    """This function merges the dataframe containing the data extracted from the raw images and from the clinical data
+
+    Parameters
+    ----------
+    imaging_data: DataFrame
+        Dataframe containing the data extracted from the raw images
+
+    df_clinical_complete: DataFrame
+        Dataframe containing the clinical data
+
+    Returns
+    -------
+    df_complete: DataFrame
+        Dataframe containing the merged data
+    """
+    df_complete = imaging_data.merge(
+        df_clinical_complete,
+        how="inner",
+        left_on=["source_id", "source_ses_id"],
+        right_on=["blinded_code", "visit"],
+    )
+    df_complete = df_complete.loc[:, ~df_complete.columns.duplicated()]
+    return df_complete
+
+
+def read_imaging_data(source_path: PathLike) -> DataFrame:
+    """This function finds the imaging data and filters it.
+
+    Parameters
+    ----------
+    source_path: PathLike
+        Path to the raw data
+
+    Returns
+    -------
+    df_dicom: DataFrame
+        Dataframe containing the data extracted.
+    """
+    return filter_dicoms(
+        pd.DataFrame(find_dicoms(source_path), columns=["source_path", "source"])
+    )
+
+
+def compute_baseline_date(df: DataFrame) -> DataFrame:
+    """Computes the baseline date by taking the minimum
+    acq_date for each subject.
+
+    Parameters
+    ----------
+    df: Dataframe
+
+    Returns
+    -------
+    Dataframe
+        Contains the baseline date.
+    """
+    df_1 = (
+        df[["source_id", "source_ses_id", "acq_date"]]
+        .groupby(["source_id", "source_ses_id"])
+        .min()
+    )
+    df_2 = df[["source_id", "acq_date"]].groupby(["source_id"]).min()
+    return df_1.join(df_2.rename(columns={"acq_date": "baseline"}))
+
+
+def compute_session_numbers(df: DataFrame) -> DataFrame:
+    """Computes the session number in months by subtracting the
+    baseline date from the acq_date and converting to months.
+
+    Parameters
+    ----------
+    df:Dataframe
+
+    Returns
+    -------
+    Dataframe
+        Contains the session number in the following format: ses-MXXX.
+    """
+    df = df.assign(
+        ses_month=lambda x: (
+            (x["acq_date"].str[:4].astype("int") - x["baseline"].str[:4].astype("int"))
```

Comment:
This seems very fragile and it is a bit hard to understand what's going on since you're taking slices of things we don't know.  I think we need to extract the function doing the computations such that we can easily test it in isolation.

I don't really know what you have in the columns `acq_date` and `baseline` (strings, timestamps ??), but usually it is better to rely on proper time data format to perform such computations.

Could you give some examples of the possible inputs and expected outputs ?

Reply:
I was thinking about something more or less like this:

```python
def compute_time_delta_month(end, start):
   """Computes the number of months between end and start datetime objects."""
    return 12 * (end.year - start.year) + (end.month - start.month)

def compute_session_numbers(df: DataFrame) -> DataFrame:
    """...doc..."""
    from datetime import datetime

    for col in ("acq_date", "baseline"):
        df[col] = df[col].apply(lambda x: datetime.strptime(x, '%Y%m%d'))
    return df.assign(
        ses_month=lambda x: compute_time_delta_month(x.acq_date, x.baseline),
        session_id=lambda x: x.ses_month.map(lambda y: f"ses-M{y:03d}")
    )
```